File: lib/dramatis/runtime/thread_pool.rb
module Dramatis; end
class Dramatis::Runtime; end

class Dramatis::Runtime::ThreadPool; end
class Dramatis::Runtime::ThreadPool::PoolThread < Thread; end

class Dramatis::Runtime::ThreadPool #:nodoc: all

  def reset soft = false
    shutdown soft
    @state = :running
  end

  def initialize
    super
    @mutex = Mutex.new
    @wait = ConditionVariable.new
    @threads = []
    @size = 0
    @state = :running
  end

  def new &block
    checkout( &block )
  end

  def length
    @mutex.synchronize do
      @threads.length
    end
  end

  def size
    @mutex.synchronize do
      @size
    end
  end

  private

  def shutdown soft
    @mutex.synchronize do
      @state = :exiting
    end
    first = true
    @mutex.synchronize do
      while first or ( not soft and @size > 0 )
        first = false
        @threads.each do |thread|
          thread.send :shutdown
        end
        while thread = @threads.pop
          thread.true_join
          @size -= 1
        end
        if not soft and @size > 0
          @state = :exit_waiting
          @wait.wait @mutex
          @state = :exiting
        end
      end
    end
  end

  def checkin thread
    @mutex.synchronize do
      @threads << thread
      if @state == :exit_waiting
        @wait.signal
      end
    end
  end

  def checkout &block
    raise "hell" if @state == :exiting
    t = nil
    @mutex.synchronize do
      if @threads.length == 0
        pt = PoolThread.new self
        @threads << pt
        @size += 1
      end
      t = @threads.pop
    end
    t.send :awaken, &block
    t
  end
  
end

class Dramatis::Runtime::ThreadPool::PoolThread < Thread #:nodoc: all

  def initialize pool
    @pool = pool
    @mutex = Mutex.new
    @wait = ConditionVariable.new
    @state = :running
    super() do
      self.abort_on_exception = true
      target
    end
  end

  alias true_join join

  def join
    # I've thought about this. It would be cool to implement join. It's not
    # too hard to do it when threads aren't reused ... which is kinda dumb.
    # It's possible to do it when threads are reused, by coding an allocation
    # counter in the "thread" object (but not the native thread). It'd be
    # cool, but I don't need it, so, oh well.
    raise "not implemented"
  end

  private

  def shutdown
    @mutex.synchronize do
      old_state, @state = @state, :exiting
      # p "#{Thread.current} shutdown #{old_state} #{@state}"
      if old_state == :waiting
        @wait.signal
      end
    end
  end

  def awaken &block
    @mutex.synchronize do
      @block = block
      old_state, @state = @state, :called
      if old_state == :waiting
        # p "#{Thread.current} signalling"
        @wait.signal
      elsif old_state == :exiting
        warn("AWAKEN AFTER EXIT")
        raise "AWAKEN AFTER EXIT"
      elsif old_state != :running
        warn("AWAKEN BAD STATE " + old_state)
        raise "AWAKEN BAD STATE " + old_state
      end
    end
  end

  def target
    while true do
      @mutex.synchronize do
        if @state == :exiting
          return
        elsif @state == :running
          @state = :waiting
          @wait.wait @mutex
          raise "hell" if @state == :waiting
        elsif @state == :called
          begin
            @block.call
          ensure
            @state = :running
            @pool.send :checkin, self
          end
        else
          warn( "!!FAIL!! " + @state)
          raise "!!FAIL!! " + @state
        end
      end
    end
  end

end

File: lib/dramatis/runtime/thread_pool.py
from __future__ import absolute_import
from __future__ import with_statement

from logging import warning

import threading
from threading import Lock
from threading import Condition
from threading import Thread
from traceback import print_exc

class ThreadPool(object):

    def reset(self, soft = False):
        self._shutdown( soft )
        self._state = "running"

    def __init__(self):
        super(ThreadPool,self).__init__()
        self._mutex = Lock()
        self._wait = Condition(self._mutex)
        self._threads = []
        self._size = 0
        self._state = "running"

    def __call__(self, target, args = []):
        return self.checkout(target, args)

    @property
    def length(self):
        with self._mutex:
            return len(self._threads)

    @property
    def size(self):
        with self._mutex:
            return self._size

    def _shutdown(self,soft):
        with self._mutex:
            self._state = "exiting"
        first = True
        with self._mutex:
            while first or ( not soft and self._size > 0 ):
                first = False
                for thread in self._threads:
                    thread._PoolThread__shutdown()
                while len(self._threads) > 0:
                    thread = self._threads.pop()
                    thread.join()
                    self._size -= 1
                if not soft and self._size > 0:
                    self._state = "exit_waiting"
                    self._wait.wait()
                    self._state = "exiting"

    def checkin(self,thread):
        with self._mutex:
            self._threads.append( thread )
            if self._state == "exit_waiting":
                self._wait.notify()

    def checkout(self, target, args):
        assert self._state != "exiting"
        t = None
        with self._mutex:
            if len(self._threads) == 0:
                pt = PoolThread(self)
                pt.setDaemon(True)
                pt.start()
                self._threads.append( pt )
                self._size += 1
            t = self._threads.pop()
        t._PoolThread__awaken( target, args )
        return t

class PoolThread(Thread):

    def __init__(self,pool):
        super(PoolThread,self).__init__( target = self.__target )
        self._pool = pool
        self.__mutex = Lock()
        self.__wait = Condition( self.__mutex )
        self.__state = "running"

    def __shutdown(self):
        with self.__mutex:
            old_state, self.__state = self.__state, "exiting"
            if old_state == "waiting":
                self.__wait.notify()

    def __awaken(self,  target, args ):
        # warning( "awaken " + str(args[0]) )
        with self.__mutex:
            self.__current_target = target
            self.__args = args
            old_state, self.__state = self.__state, "called"
            if old_state == "waiting":
                self.__wait.notify()
            elif old_state == "exiting":
                warning("AWAKEN AFTER EXIT")
                raise "AWAKEN AFTER EXIT"
            elif old_state != "running":
                warning("AWAKEN BAD STATE " + old_state)
                raise ("AWAKEN BAD STATE " + old_state)

    def __target( self ):
        while True:
            with self.__mutex:
                # warning(str(self) + " llop start " + self.__state)
                if self.__state == "exiting":
                    return
                elif self.__state == "running":
                    self.__state = "waiting"
                    # warning(str(self) + " sleeping")
                    self.__wait.wait()
                    # warning(str(self) + " waking up")
                    assert self.__state != "waiting"
                elif self.__state == "called":
                    try:
                        # warning(self.__current_target)
                        # warning(self.__args)
                        # warning(str(self) + " starting " + str(self.__args[0]))
                        self.__current_target(*self.__args)
                        # warning(str(self)+" done")
                    except:
                        print_exc()
                        raise
                    finally:
                        self.__state = "running"
                        self._pool.checkin( self )
                else:
                    warning ( "!!FAIL!! " + self.__state)
                    raise "!!FAIL!! " + self.__state


